Add `CommandHadNoEffect` and `ImpossibleRevokeCommand` notifications (#54)
neo4j/docs-status-codes: modules/ROOT/pages/notifications/all-notifications.adoc

m|DEPRECATION
|===

[#_security_notifications]
== `SECURITY` category

Notifications in the security category indicates that the result of the query or command might have a potential security issue.
Make sure that the behaviour is what was intended.

[#_neo_clientnotification_security_commandhadnoeffect]
=== Neo.ClientNotification.Security.CommandHadNoEffect

.Notification details
[cols="<1s,<4"]
|===
|Code
m|Neo.ClientNotification.Security.CommandHadNoEffect
|Title

> a|`<command>` has no effect.

Author: How do we want to handle details in the titles? I added `Full title of the returned code` in the examples, is that a good way?

Reviewer: I think that works, but would like to hear what @renetapopova thinks as well

|Severity
m|INFORMATION
|Category
m|SECURITY
|===

The `<command>` in the title is the part of the command which had no effect.
This could be either the full command given by the user or a subset of the given command.

> .Granting a role to a user who already have that role

Author: ```suggestion
.Granting a role to a user who already has that role
```
I think this was the only one you added a . to

> .Granting a role to a user who already have that role

Reviewer 2: oh, that was by mistake

The role already has the privilege. See Status Codes documentation for more information.

Suggestions for improvement::
Make sure this was the intended privilege and role.
====

.Revoking a privilege from a role that does not have that privilege
====
Command::
+
[source, cypher]
----
REVOKE WRITE ON GRAPH * FROM editor
----

Full title of the returned code::
`REVOKE DENY WRITE ON GRAPH * FROM editor` has no effect.

Description of the returned code::
The role does not have the privilege. See Status Codes documentation for more information.

Suggestions for improvement::
Make sure this was the intended privilege and role.
====

[#_neo_clientnotification_security_impossiblerevokecommand]
=== Neo.ClientNotification.Security.ImpossibleRevokeCommand

.Notification details
[cols="<1s,<4"]
|===
|Code
m|Neo.ClientNotification.Security.ImpossibleRevokeCommand
|Title
a|`<command>` has no effect.
|Severity
m|WARNING
|Category
m|SECURITY
|===

The `<command>` in the title is the part of the command which had no effect.
This could be either the full command given by the user or a subset of the given command.

.Revoking a non-existing role from a user
====
Command::
+
[source, cypher]
----

> REVOKE ROLE Foo, reader FROM neo4j

Reviewer 2: Here `Foo `is the role and `neo4j` is the user. Perhaps, we could stick to the same names.

Author: Not sure I follow, but I can look over the names again

Reviewer 2: I mean that in the examples, sometimes Foo is a user, sometimes a role, and sometimes a graph.

Author: I think Foo is often the one that is non-existing which was the pattern I was trying to follow but I'll look it over next week hopefully

Reviewer 3: fixed (https://github.com/neo4j/docs-status-codes/pull/54#discussion_r1193752367)

----

Full title of the returned code::
`REVOKE ROLE Foo FROM neo4j` has no effect.

Description of the returned code::
Role does not exist. Make sure nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::

> Make sure nothing is misspelled and that this was the intended role and user.

Reviewer: Maybe a simpler suggestion would be just something like
'Make sure that the role was not misspelled or another role was intended.' You don't have to check if it was the intended user if the role does not exists. 

Similar for the next example

Command::
+
[source, cypher]
----
REVOKE ROLE reader FROM Foo
----

Full title of the returned code::
`REVOKE ROLE reader FROM Foo` has no effect.

Description of the returned code::
User does not exist. Make sure nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Make sure nothing is misspelled and that this was the intended role and user.
====

.Revoking a privilege from a non-existing role
====
Command::
+
[source, cypher]
----
REVOKE GRANT WRITE ON GRAPH * FROM Foo
----

Full title of the returned code::
`REVOKE GRANT WRITE ON GRAPH * FROM Foo` has no effect.

Description of the returned code::
Role does not exist. Make sure nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Make sure nothing is misspelled and that this was the intended role.
====

.Revoking a privilege on a non-existing graph/database from a role
====
Command::
+
[source, cypher]
----
REVOKE GRANT WRITE ON GRAPH Foo FROM editor
----

Full title of the returned code::

> `REVOKE GRANT WRITE ON GRAPH Foo FROM editor` has no effect.

Reviewer 3: fixed (https://github.com/neo4j/docs-status-codes/pull/54#discussion_r1193752367)

Description of the returned code::

> Database 'Foo' does not exist. Make sure nothing is misspelled.

Author: The message we currently have for the error on grant/deny have ' and not ` so I'd keep those here
```suggestion
Graph 'Foo' does not exist. Make sure nothing is misspelled.
```

> Database 'Foo' does not exist. Make sure nothing is misspelled.

Reviewer 3: I've updated the names to:
- Existing users: `john`, `jane`
- Missing user: `alice`
- Existing roles: `reader`, `admin`, `PUBLIC`
- Missing roles: `manager`
- Existing database/graph: `neo4j`
- Missing database/graph: `neo3j`

I kept graph and database name the same because [that is how we refer to graphs](https://neo4j.com/docs/operations-manual/current/manage-databases/introduction/#manage-databases-concepts):
> Many administrative commands refer to a specific graph by using the database name.

This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::

> Make sure nothing is misspelled and that this was the intended graph/database.

Author: then I would want to add another example with database as well since it can happen for moth graph and database privileges

====

.Revoking a privilege from a role with wildcard graph/database parameter
====
Parameter::
+
[source, javascript]
----
{
    "database": "*"
}
----
Command::
+
[source, cypher]
----
REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC
----

Full title of the returned code::
`REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC` has no effect.

Description of the returned code::

> Parameterized database and graph names do not support wildcards. Make sure nothing is misspelled.

Author: ```suggestion
Parameterized database names do not support wildcards.
Verify nothing is misspelled.
```

> Parameterized database and graph names do not support wildcards. Make sure nothing is misspelled.

Author: Actually the current error for grant/deny with this cause already says:
```
Parameterized database and graph names do not support wildcards.
```
so I say we keep the same message here (I'm also unsure if we know which of them we have when it's thrown)

```suggestion
Parameterized database and graph names do not support wildcards.
Verify nothing is misspelled.
```

> Parameterized database and graph names do not support wildcards. Make sure nothing is misspelled.

Reviewer 3: I split the graph & database examples, and updated description / suggested improvement accordingly. But left the line that Therese highlighted intact.

m|DEPRECATION
|===

[#_security_notifications]
== `SECURITY` category

Security notifications indicate that the result of the query or command might have a potential security issue.
Verify that this is the intended behavior of your query or command.

[#_neo_clientnotification_security_commandhadnoeffect]

> === Neo.ClientNotification.Security.CommandHadNoEffect

Author: Just as a reminder, if we update this we should update the CIP and regular PR as well since it's the code part

.Notification details
[cols="<1s,<4"]
|===
|Code
m|Neo.ClientNotification.Security.CommandHadNoEffect
|Title

> a|`<command>` has no effect.

Author: what does this do?

> a|`<command>` has no effect.

Reviewer 2: Doesn't do anything. It just shows that there is a note below the table.

}
----
Command::
+
[source, cypher]
----
REVOKE GRANT CREATE ON GRAPH $graph FROM PUBLIC
----

Full title of the returned code::
`REVOKE GRANT CREATE ON GRAPH $graph FROM PUBLIC` has no effect.

Description of the returned code::
Parameterized database and graph names do not support wildcards. Verify that nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Use `GRAPH *` without parameter to revoke the privilege on all databases.
====

.Revoking a privilege from a role with wildcard database parameter
====
Parameter::
+
[source, javascript]
----
{
    "database": "*"
}
----
Command::
+
[source, cypher]
----
REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC
----

Full title of the returned code::
`REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC` has no effect.

Description of the returned code::
Parameterized database and graph names do not support wildcards. Verify that nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::

> Use `DATABASE *` without parameter to revoke the privilege on all databases.

Reviewer 2: ```suggestion
Use `DATABASE *` without the parameter to revoke the privilege on all databases.
```

See Status Codes documentation for more information.

Suggestions for improvement::
Verify that this is the intended graph and that it is spelled correctly.
====

.Revoking a privilege on a non-existing database from a role
====
Command::
+
[source, cypher]
----
REVOKE GRANT ACCESS ON DATABASE neo3j FROM editor
----

Full title of the returned code::
`REVOKE GRANT ACCESS ON DATABASE neo3j FROM editor` has no effect.

Description of the returned code::
Database 'neo3j' does not exist. Make sure nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Verify that this is the intended database and that it is spelled correctly.
====

.Revoking a privilege from a role with wildcard graph parameter
====
Parameter::
+
[source, javascript]
----
{
    "graph": "*"
}
----
Command::
+
[source, cypher]
----
REVOKE GRANT CREATE ON GRAPH $graph FROM PUBLIC
----

Full title of the returned code::
`REVOKE GRANT CREATE ON GRAPH $graph FROM PUBLIC` has no effect.

Description of the returned code::
Parameterized database and graph names do not support wildcards. 

> Verify that everything is spelled correctly.

Author: I think this sentence have been updated by mistake, should be
```suggestion
Make sure nothing is misspelled.
```
to match the other ones (it's just the cause that differs)

Reviewer 2: I _think_ writing positively is better when possible.

Reviewer 3: Another factor is that we show description and suggestion for improvement next to each other in the docs, so we get repeated sentences. Using `Make sure` in description and `Verify that` in suggestion for improvement makes the text a little more natural.

This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::

> Use `GRAPH *` without the parameter to revoke the privilege on all databases.

Author: ```suggestion
Use `GRAPH *` without the parameter to revoke the privilege on all graphs.
```
Since this is just the suggestion text (and graph example)

====

.Revoking a privilege from a role with wildcard database parameter
====
Parameter::
+
[source, javascript]
----
{
    "database": "*"
}
----
Command::
+
[source, cypher]
----
REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC
----

Full title of the returned code::
`REVOKE GRANT ACCESS ON DATABASE $database FROM PUBLIC` has no effect.

Description of the returned code::
Parameterized database and graph names do not support wildcards. 

> Verify that everything is spelled correctly.

Author: I think this sentence have been updated by mistake, should be
```suggestion
Make sure nothing is misspelled.
```
to match the other ones (it's just the cause that differs)

+
[source, cypher]
----
REVOKE GRANT WRITE ON GRAPH * FROM manager
----

Full title of the returned code::
`REVOKE GRANT WRITE ON GRAPH * FROM manager` has no effect.

Description of the returned code::
Role does not exist. Make sure nothing is misspelled.
This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Verify that this is the intended role and that it was not misspelled.
====

.Revoking a privilege on a non-existing graph from a role
====
Command::
+
[source, cypher]
----
REVOKE GRANT WRITE ON GRAPH neo3j FROM editor
----

Full title of the returned code::
`REVOKE GRANT WRITE ON GRAPH neo3j FROM editor` has no effect.

Description of the returned code::

> Graph 'neo3j' does not exist. Make sure nothing is misspelled.

Author: I don't think this one would change either. The current error for grant/deny says: `Database 'X' does not exist.` for both graph and database

This notification will become an error in a future major version.
See Status Codes documentation for more information.

Suggestions for improvement::
Verify that this is the intended graph and that it was not misspelled.
====

.Revoking a privilege on a non-existing database from a role
====
Command::
+
[source, cypher]
----

> REVOKE GRANT WRITE ON GRAPH neo3j FROM editor

Author: should have a database privilege not a graph privilege
